### PR TITLE
fix(cxx_indexer) the memory of the old client is released if it's own…

### PR DESCRIPTION
**Pull request by leonzhuLGU**
…ed by the diags, and cannot be recovered by the scopeguard

### CODE REVIEW

File: kythe/cxx/indexer/cxx/marked_source.cc [hunk 1]
```diff
@@ -596,9 +596,10 @@ void MarkedSourceGenerator::ReplaceMarkedSourceWithTemplateArgumentList(
   // diagnostics from the typechecker.
   clang::IgnoringDiagConsumer consumer;
   auto* diags = &cache_->source_manager().getDiagnostics();
-  auto guard = MakeScopeGuard(
-      [diags = diags, client = diags->getClient(),
-       owned = diags->ownsClient()] { diags->setClient(client, owned); });
+  
+  auto OwningPreviousClient = diags->takeClient();
+  auto PreviousClient = diags->getClient();
```
Comment (shahms) [inline kythe/cxx/indexer/cxx/marked_source.cc:601]: please use `lower_with_under` for local variables here

File: kythe/cxx/indexer/cxx/marked_source.cc [hunk 2]
```diff
@@ -697,6 +698,8 @@ void MarkedSourceGenerator::ReplaceMarkedSourceWithTemplateArgumentList(
     }
     *next_arg->mutable_pre_text() = pre_text;
   }
+
+  diags->setClient(PreviousClient, !!OwningPreviousClient.release());
```
Comment (shahms) [inline kythe/cxx/indexer/cxx/marked_source.cc:702]: This should still be done with a scope guard:
```
auto guard = MakeScopeGuard([diags = diags, client = diags->getClient(), owned = diags->takeClient()] {
  diags->setClient(client, owned.release() != nullptr));
});
```

**Review by shahms — COMMENTED** [review]
Thanks! A few comments, but I appreciate the PR.

File: kythe/cxx/indexer/cxx/marked_source.cc (hunk 2 above)
Comment (leonzhuLGU) [reply kythe/cxx/indexer/cxx/marked_source.cc]: I think the lambda expression will capture the smart pointer as const, therefore you cannot call release() inside the body without some proper cast.

Comment (shahms) [reply kythe/cxx/indexer/cxx/marked_source.cc]: Fair point, but you should just need to make the lambda mutable:
```
auto guard = MakeScopeGuard([diags = diags, client = diags->getClient(), owned = diags->takeClient()] () mutable {
  diags->setClient(client, owned.release() != nullptr));
});
```

File: kythe/cxx/indexer/cxx/marked_source.cc [hunk 3]
```diff
@@ -697,6 +698,8 @@ void MarkedSourceGenerator::ReplaceMarkedSourceWithTemplateArgumentList(
     }
     *next_arg->mutable_pre_text() = pre_text;
   }
+
+  diags->setClient(previous_client, !!owning_previous_client.release());
```
Comment (shahms) [inline kythe/cxx/indexer/cxx/marked_source.cc:702]: please use an explicit comparison to nullptr

File: kythe/cxx/indexer/cxx/marked_source.cc (hunk 2 above)
Comment (leonzhuLGU) [reply kythe/cxx/indexer/cxx/marked_source.cc]: I see, I have changed to use scopeguard which looks more elegant

**Review by shahms — APPROVED** [review]
Thanks a lot!

**Pull request merged**